feat: add voice tracing and evaluation docs scaffold (#4368)
langchain-ai/docs: src/langsmith/trace-openai-realtime.mdx

---
title: Trace OpenAI Realtime applications
sidebarTitle: OpenAI Realtime
description: Trace OpenAI Realtime voice agents in LangSmith using the LangSmith SDK.
---

The [OpenAI Realtime API](https://platform.openai.com/docs/guides/realtime) powers low-latency speech-to-speech voice agents over a persistent WebSocket connection. This guide shows how to trace a Realtime app to LangSmith.

Before you start, read [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals). OpenAI Realtime is a [speech-to-speech](/langsmith/trace-voice-fundamentals#speech-to-speech-s2s) architecture, so this guide applies the conventions for that shape: one conversation is one trace, you trace the **event payloads** that cross the wire (not turns), and you attach a single combined audio recording to the root run with `ls_modality="audio"`.

For a complete implementation, see the [demo repository](https://github.com/langchain-ai/voice-agents-tracing).

## Why you build the trace yourself

> OpenAI Realtime has no telemetry of its own. It is a WebSocket connection (`client.realtime.connect(model=...)` in the `openai` Python SDK) carrying a bidirectional stream of typed JSON events — there are no OpenTelemetry spans, callbacks, or tracing hooks. To get a trace, you observe the event stream and build the trace yourself with the LangSmith SDK. (This is the opposite of [LiveKit](/langsmith/trace-with-livekit) and [Pipecat](/langsmith/trace-with-pipecat), which run in-process and emit OTel spans you translate.)

Reviewer: ```suggestion
OpenAI Realtime has no telemetry of its own. It is a WebSocket connection (`client.realtime.connect(model=...)` in the `openai` Python SDK) carrying a bidirectional stream of typed JSON events. There are no OpenTelemetry spans, callbacks, or tracing hooks. To get a trace, you observe the event stream and build the trace yourself with the LangSmith SDK. (This is the opposite of [LiveKit](/langsmith/trace-with-livekit) and [Pipecat](/langsmith/trace-with-pipecat), which run in-process and emit OTel spans you translate.)
```

Every event has a discriminated `type` string, so dispatch is a simple match on `event.type`.

## The event model

> The client **sends** events to configure the session and stream audio. You don't trace these as spans — they're requests, and their effects come back as server events:

Reviewer: ```suggestion
The client **sends** events to configure the session and stream audio. You don't trace these as spans, they're requests, and their effects come back as server events:
```

| Client event | What it does |
| --- | --- |
| `session.update` | Configure the session: instructions, voice, audio formats, transcription model, turn detection, tools. |
| `input_audio_buffer.append` | Stream a base64 PCM16 mic chunk. Sent continuously. |

> | `conversation.item.create` | Add an item — used to return a `function_call_output` after running a tool. |

Reviewer: ```suggestion
| `conversation.item.create` | Add an item—used to return a `function_call_output` after running a tool. |
```

| `response.create` | Ask the model to generate a response (needed explicitly when turn detection uses `create_response: false`). |

The server **sends back** events, which you trace. Roughly in per-turn order:

| Server event | What it carries | Traced? |
| --- | --- | --- |
| `session.created` / `session.updated` | Handshake / config acknowledgement. | Yes |

> | `input_audio_buffer.speech_started` | Server VAD heard the user start — the **barge-in signal**; flush the speaker buffer. | Yes |

Reviewer: ```suggestion
| `input_audio_buffer.speech_started` | Server VAD heard the user start—the **barge-in signal**; flush the speaker buffer. | Yes |
```

| `input_audio_buffer.speech_stopped` | Server VAD heard the user stop. | Yes |
| `input_audio_buffer.committed` | The audio buffer became a conversation item. | Yes |
| `conversation.item.created` | An item was added server-side. | Yes |
| `conversation.item.input_audio_transcription.completed` | The full user transcript for the turn. | Yes |
| `response.created` | The model started generating. | Yes |

> | `response.output_audio.delta` | One chunk of agent speech (base64 PCM16); hundreds per response. | No — played, never spanned |

Reviewer: ```suggestion
| `response.output_audio.delta` | One chunk of agent speech (base64 PCM16); hundreds per response. | No—played, never spanned |
```

| `response.done` | The complete response object: all output items (including every `function_call`), plus token usage. | Yes |
| `error` | Server-reported error. | Yes |
| `rate_limits.updated` | Quota bookkeeping. | Yes |

<Note>
**The noise rule:** skip every event type ending in `.delta`, because the matching `.done` event repeats the complete payload — tracing both records everything twice. `response.output_audio.delta` in particular is the agent's voice: hundreds of chunks per response that would bury the trace. Play it to the speaker, but never make it a span.
</Note>

## Build the trace your way

> LangSmith doesn't ship a dedicated OpenAI Realtime SDK integration, so you build the trace yourself from the event stream. The upside is full control: you decide which events become spans, how they're named, and what lands in inputs versus outputs. This guide walks through an approach that works well, and you can adapt it to whatever shape you want.

Author: remove

> **The noise rule:** skip every event type ending in `.delta`, because the matching `.done` event repeats the complete payload — tracing both records everything twice. `response.output_audio.delta` in particular is the agent's voice: hundreds of chunks per response that would bury the trace. Play it to the speaker, but never make it a span.

Reviewer: ```suggestion
**The noise rule:** we recommend skipping every event type ending in `.delta`, because the matching `.done` event repeats the complete payload. Tracing both records everything twice. `response.output_audio.delta` in particular is the agent's voice: hundreds of chunks per response that would bury the trace. Play it to the speaker, but never make it a span.
```

## How events map to LangSmith runs

> One conversation is one trace, with a span per traced event in arrival order:

Reviewer: ```suggestion
We recommend tracing the whole conversation as a single trace, with one span per traced event.
```

│   attachments: conversation.wav (stereo: L=user, R=agent)
│
├─ input_audio_buffer.speech_started
├─ input_audio_buffer.speech_stopped
├─ conversation.item.input_audio_transcription.completed
├─ response.created
├─ response.function_call_arguments.done
├─ response.done
│   └─ lookup_weather × N                         ← real tool runs, nested (see below)
├─ response.done                                  ← the spoken follow-up after tools
└─ error                                          ← only if the server sent one
```

- **A span per event, named by `event.type`, in arrival order.** The trace mirrors exactly what the server sent.
- **Direction decides inputs vs outputs.** Events the user sent toward the model (`input_audio_buffer.*`, the transcription completion) put their payload in the span's `inputs`; everything the model or server sent back (`response.*`, `error`) goes in `outputs`.
- **Scrub every payload.** Replace raw `bytes` (audio, base64 blobs) with a `"<N bytes>"` placeholder and truncate long strings, so transcripts, tool args, configs, and usage land intact but megabytes of audio never ship to LangSmith.

> - **The root run** opens at connect and is finalized in a `finally` block. It carries roll-up metadata (`event_count`, `duration_s`) and one combined stereo WAV attachment — left channel the user mic, right channel the agent speech recorded at the *speaker*, so audio flushed by a barge-in is never in the recording.

Reviewer: ```suggestion
- **The root run** opens at connect and is finalized in a `finally` block. It carries roll-up metadata (`event_count`, `duration_s`) and one combined stereo WAV attachment—left channel the user mic, right channel the agent speech recorded at the *speaker*, so audio flushed by a barge-in is never in the recording.
```

> - **Errors** are stamped on the root run before finalizing, so a dead session never looks like a clean exit.

Reviewer: ```suggestion
```

### Tool calls become real nested runs

With Realtime, tool execution is the client's job: the server announces a `function_call`, your process runs the function, sends back a `function_call_output` item, and requests the follow-up response. That execution is real work in your process, so trace it as a child run nested inside the announcing event's span. (Contrast with [Gemini Live via ADK](/langsmith/trace-gemini-live), where the framework runs the tool and you only observe `function_call` / `function_response` events.)

Two practical details:

- Read the tool calls from **`response.done`'s payload** (`response.output` items with `type == "function_call"`), not from the earlier `response.function_call_arguments.done` events. You must wait for `response.done` anyway, since the API allows only one active response at a time.

> - Send `function_call_output` and request the follow-up `response.create()` from inside the `response.done` handler.

Reviewer: ```suggestion
```

## Installation

```bash
pip install "langsmith>=0.4" "openai>=1.50"
```

The demo also uses `sounddevice` and `numpy` for the mic/speaker and to build the WAV attachment.

## Set up your environment

> This path uses the LangSmith SDK directly — no OpenTelemetry or OTLP setup.

Reviewer: ```suggestion
The following steps demonstrate how to trace using our SDK. You can also trace using OpenTelemetry directly.
```

```bash
export LANGSMITH_API_KEY=...
export LANGSMITH_TRACING=true
export LANGSMITH_PROJECT=my-voice-app
export OPENAI_API_KEY=...
```

## Quickstart

### Step 1: Open the conversation root at connect time

Use one `RunTree` per conversation:

```python
from langsmith import RunTree

root = RunTree(
    name="realtime_session",
    run_type="chain",
    inputs={},
    project_name="my-voice-app",
    extra={"metadata": {"thread_id": thread_id, "model": MODEL, "ls_modality": "audio"}},
)
root.post()
```

A stable `thread_id` you generate per conversation (for example, a UUID) groups the trace into a LangSmith [thread](/langsmith/threads); `ls_modality="audio"` marks it as a voice conversation.

### Step 2: Span each received event, skipping the noise

```python
async for event in connection:
    if event.type == "response.output_audio.delta":
        speaker.write(base64.b64decode(event.delta))   # play it…
        continue                                        # …never span it
    if event.type.endswith(".delta"):

>         continue                                        # .done repeats the payload

Author: Simplify to just: if event type ends with delta, continue; otherwise trace the event. In other words, let's just focus on tracing, or you can put.. to show the inline code. Let's not worry about speaker, because speaker is like an implementation detail for a demo, but other people might have different variable names or different ways or different places.

>     with session.event_span(event, session.now(),

Author: Event span isn't provided anywhere, so if we're using event span, we need to provide a defined event span function at the top just to show what it is and what it does.

                            name=event.type,
                            inbound=is_inbound(event.type)) as event_run:
        ...   # your handling for this event type
```

`event_span` creates the child run (payload scrubbed, routed to inputs or outputs by `inbound`), posts it, and ends it when the handler body exits — so any work done inside the block nests under the event.

### Step 3: Run tools nested under the announcing event

```python
from langsmith.run_helpers import tracing_context

if event.type == "response.done":
    calls = [i for i in (event.response.output or []) if i.type == "function_call"]
    for call in calls:
        with tracing_context(parent=event_run):
            result = await execute_tool(call.name, call.arguments)  # traced child
        await connection.conversation.item.create(item={
            "type": "function_call_output",
            "call_id": call.call_id,
            "output": json.dumps(result),
        })
    if calls:
        await connection.response.create()   # ask for the spoken follow-up
```

### Step 4: Finalize in a `finally` block

Stamp any error on the root run, attach the WAV, then end and patch the root.

> ### Gotchas

Reviewer: Same comment here.

- **`create_response: false` requires a manual `response.create()`.** If your turn-detection config disables auto-response (to wait for the user transcript before replying), you must send `response.create()` yourself or the agent never answers.

> - **One active response at a time.** Don't call `response.create()` until the prior `response.done` arrives — this is why tool outputs and the follow-up request happen inside the `response.done` handler.

Reviewer: ```suggestion
- **One active response at a time.** Don't call `response.create()` until the prior `response.done` arrives. This is why tool outputs and the follow-up request happen inside the `response.done` handler.
```

- **Always scrub before attaching payloads.** Forgetting this ships base64 audio strings to LangSmith and makes runs enormous.

> - **Finalize in `finally` and set the root error on exceptions,** or crashes produce traces that look like clean sessions — or no root at all.

Reviewer: ```suggestion
- **Finalize in `finally` and set the root error on exceptions,** or crashes produce traces that look like clean sessions, or no root at all.
```

> - **Record audio at the playout boundary.** Agent audio recorded at receipt includes speech the user never heard (barge-in truncation). Record what was played to the speaker instead.

Reviewer: ```suggestion
```

## Next steps

<CardGroup cols={2}>
  <Card title="Voice fundamentals" icon="waveform" href="/langsmith/trace-voice-fundamentals">
    Core conventions for tracing voice agents.
  </Card>
  <Card title="Upload files with traces" icon="paperclip" href="/langsmith/upload-files-with-traces">
    Attach the conversation audio recording to your trace.

>   </Card>

Author: Can we add a section on uploading audio recordings to traces just for voice traces, so we can talk about how everything should be one file and how it should be recorded as close to the client as possible and how to do it for the different model providers?

---
title: Trace OpenAI Realtime applications
sidebarTitle: OpenAI Realtime
description: Trace OpenAI Realtime voice agents in LangSmith using the LangSmith SDK.
---

The [OpenAI Realtime API](https://platform.openai.com/docs/guides/realtime) powers low-latency speech-to-speech voice agents. This guide shows how to trace a Realtime app to LangSmith.

> OpenAI Realtime is a speech-to-speech model: it processes audio natively and exchanges a continuous stream of typed JSON events with your application over a persistent WebSocket connection, rather than making discrete request/response calls. The sections below walk through those events and how to turn them into a LangSmith trace. For our high-level philosophy on getting the most out of your voice agent traces, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).

Reviewer: ```suggestion
OpenAI Realtime is a speech-to-speech model: it processes audio natively and exchanges a continuous stream of typed JSON events with your application over a persistent WebSocket connection, rather than making discrete request/response calls. The following sections show those events and how to turn them into a LangSmith trace. For our high-level philosophy on getting the most out of your voice agent traces, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).
```

---
title: Trace OpenAI Realtime applications
sidebarTitle: OpenAI Realtime
description: Trace OpenAI Realtime voice agents in LangSmith using the LangSmith SDK.
---

The [OpenAI Realtime API](https://platform.openai.com/docs/guides/realtime) powers low-latency speech-to-speech voice agents. This guide shows how to trace a Realtime app to LangSmith.

> OpenAI Realtime is a speech-to-speech model: it processes audio natively and exchanges a continuous stream of typed JSON events with your application over a persistent WebSocket connection (`client.realtime.connect(model=...)` in the `openai` Python SDK), rather than making discrete request/response calls. The sections below walk through those events and how to turn them into a LangSmith trace. For our high-level philosophy on getting the most out of your voice agent traces, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).

Author: ```suggestion
OpenAI Realtime is a speech-to-speech model: it processes audio natively and exchanges a continuous stream of typed JSON events with your application over a persistent WebSocket connection, rather than making discrete request/response calls. The sections below walk through those events and how to turn them into a LangSmith trace. For our high-level philosophy on getting the most out of your voice agent traces, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).
```

| `input_audio_buffer.speech_stopped` | Server VAD heard the user stop. | Yes |
| `input_audio_buffer.committed` | The audio buffer became a conversation item. | Yes |
| `conversation.item.created` | An item was added server-side. | Yes |
| `conversation.item.input_audio_transcription.completed` | The full user transcript for the turn. | Yes |
| `response.created` | The model started generating. | Yes |
| `response.output_audio.delta` | One chunk of agent speech (base64 PCM16); hundreds per response. | No — played, never spanned |
| `response.output_audio_transcript.delta` | Streaming fragment of the agent's transcript. | No |
| `response.output_audio_transcript.done` | The agent's full transcript for the response. | Yes |
| `response.function_call_arguments.delta` / `.done` | Streaming / final tool-call arguments. | `.done` only |
| `response.output_item.*`, `response.content_part.*` | Structural progress of the response. | Yes |
| `response.done` | The complete response object: all output items (including every `function_call`), plus token usage. | Yes |
| `error` | Server-reported error. | Yes |
| `rate_limits.updated` | Quota bookkeeping. | Yes |

<Note>

> **The noise rule:** skip every event type ending in `.delta`, because the matching `.done` event repeats the complete payload — tracing both records everything twice. `response.output_audio.delta` in particular is the agent's voice: hundreds of chunks per response that would bury the trace. Play it to the speaker, but never make it a span.

Author: phrase as we recommend... - never tell the user what to do, sounds condescending

</Note>

## How events map to LangSmith runs

> One conversation is one trace, with a span per traced event in arrival order:

Author: You should trace the conversation....

├─ response.done                                  ← the spoken follow-up after tools
└─ error                                          ← only if the server sent one
```

- **A span per event, named by `event.type`, in arrival order.** The trace mirrors exactly what the server sent.
- **Direction decides inputs vs outputs.** Events the user sent toward the model (`input_audio_buffer.*`, the transcription completion) put their payload in the span's `inputs`; everything the model or server sent back (`response.*`, `error`) goes in `outputs`.
- **Scrub every payload.** Replace raw `bytes` (audio, base64 blobs) with a `"<N bytes>"` placeholder and truncate long strings, so transcripts, tool args, configs, and usage land intact but megabytes of audio never ship to LangSmith.
- **The root run** opens at connect and is finalized in a `finally` block. It carries roll-up metadata (`event_count`, `duration_s`) and one combined stereo WAV attachment — left channel the user mic, right channel the agent speech recorded at the *speaker*, so audio flushed by a barge-in is never in the recording. For the attachment API itself, see [Upload files with traces](/langsmith/upload-files-with-traces).

> - **Errors** are stamped on the root run before finalizing, so a dead session never looks like a clean exit.

Author: remove

### Tool calls become real nested runs

> With Realtime, tool execution is the client's job: the server announces a `function_call`, your process runs the function, sends back a `function_call_output` item, and requests the follow-up response. That execution is real work in your process, so trace it as a child run nested inside the announcing event's span.

Author: remove

Two practical details:

- Read the tool calls from **`response.done`'s payload** (`response.output` items with `type == "function_call"`), not from the earlier `response.function_call_arguments.done` events. You must wait for `response.done` anyway, since the API allows only one active response at a time.

> - Send `function_call_output` and request the follow-up `response.create()` from inside the `response.done` handler.

Author: remove

## Installation

```bash
pip install "langsmith>=0.4" "openai>=1.50"
```

The demo also uses `sounddevice` and `numpy` for the mic/speaker and to build the WAV attachment.

## Set up your environment

> This path uses the LangSmith SDK directly — no OpenTelemetry or OTLP setup.

Author: same as for gemini comment please

        await connection.conversation.item.create(item={
            "type": "function_call_output",
            "call_id": call.call_id,
            "output": json.dumps(result),
        })
    if calls:
        await connection.response.create()   # ask for the spoken follow-up
```

### Step 4: Finalize in a `finally` block

Stamp any error on the root run, attach the WAV, then end and patch the root.

### Gotchas

- **`create_response: false` requires a manual `response.create()`.** If your turn-detection config disables auto-response (to wait for the user transcript before replying), you must send `response.create()` yourself or the agent never answers.
- **One active response at a time.** Don't call `response.create()` until the prior `response.done` arrives — this is why tool outputs and the follow-up request happen inside the `response.done` handler.
- **Always scrub before attaching payloads.** Forgetting this ships base64 audio strings to LangSmith and makes runs enormous.
- **Finalize in `finally` and set the root error on exceptions,** or crashes produce traces that look like clean sessions — or no root at all.

> - **Record audio at the playout boundary.** Agent audio recorded at receipt includes speech the user never heard (barge-in truncation). Record what was played to the speaker instead.

Author: remove

## Attach the conversation audio

Attach a single recording of the whole conversation to the root run so you can listen to it alongside the transcript. Two things make the recording faithful to what actually happened:

- **One combined file.** Record both the user and the agent in a single file (for example, a stereo WAV with the user mic on one channel and the agent on the other) rather than separate per-event clips.
- **Recorded as close to the client as possible.** Capture the agent's audio at the playout boundary — what was sent to the speaker — not as it arrives off the wire. This way the recording reflects what the user actually heard, including speech cut off by a barge-in.

The Realtime API streams agent audio as `response.output_audio.delta` events; decode and write those bytes to your output device, and tap that same output to build the recording. Other providers stream audio differently, but the principle is the same: record one combined file at the client.

> Attach the finished file to the root run with the [attachments API](/langsmith/upload-files-with-traces). For the cross-provider rationale, see [Record a single combined audio file](/langsmith/trace-voice-fundamentals#record-a-single-combined-audio-file).

Author: show code snippet too please
